### DISCUSSION TITLE
typeof(this): A better way to specify the current declaring-type as a generic type parameter

### DISCUSSION THREAD

**Discussion by Timo-Weike** [dotnet/csharplang#5509] [Language Ideas]
# The Problem

We sometimes want to specify the current type as a type parameter for some other type.
E.g. in case of injecting a logger we get something like
~~~ cs
public class Foo
{
	public Foo(ILogger<Foo> logger)
	{
		...
	}
}
~~~
so the constructor of `Foo` wants a logger for itself and we have to explicitly state that we want a `ILogger<Foo>`.

Another example is the `IComparable<T>` or `IEquatable<T>`
~~~ cs
public Foo : IComparable<Foo>, IEquatable<Foo> { ... }
~~~
here we also have to explictly state that our class `Foo` is comparable with `Foo` objects

## Current work-arounds
One work-around I often see is this
~~~ cs
private static readonly ILog Logger = LogManager.GetLogger(MethodBase.GetCurrentMethod().DeclaringType); 
~~~
were the logger is static and the call-stack is use to determine the declaring-type

# Proposal/Idea
We could introduce some key word (or reuse an existing) that, in the context of type-parameter means the current declaring-type,
For example `typeof(this)` could be it.
Then the previous examples would become
~~~ cs
public class Foo
{
	public Foo(ILogger<typeof(this)> logger)
	{
	}
}
~~~
~~~ cs
public Foo : IComparable<typeof(this)>, IEquatable<typeof(this)> { ... }
~~~

This should work everywhere the nearest declaring-type is clear, so even in a static method and neasted classes.

~~~ cs
public class Foo
{
	class Bar : IComparable<typeof(this)> { ... }
}
~~~
here the `typeof(this)` would be resolved to `Bar`

## What about delagate, `Func` or `Action`
~~~ cs
public class Foo
{
	public void DoStuff()
	{
		Action<string> logger = (msg) => LogManager.GetLogger<typeof(this)>().Log(msg); 
	}
}
~~~
Even in this case in my opinion should `typeof(this)` be resolved to `Foo`

**Comment by jnm2** [comment]
https://github.com/dotnet/csharplang/issues/5413 is an alternative proposal that is championed and should cover these use cases.

**Reply by Timo-Weike** [reply]
The proposal directly states that what I'm proposing is not in the scope of the proposed feature

> [The self-constrained type argument must still be specified at the use site. That is the user must still fully specify `struct Half : IAdditiveIdentity<Half>` and could not choose to elide or otherwise not specify `<Half>`.](https://github.com/dotnet/csharplang/blob/main/proposals/self-constraint.md#detailed-design)

This feature would also not be applicable to the `ILogger<T>` example, because if I want to use a logger for the class `Foo` and the `ILogger<T>`  would constraint `T` with this Self-constraint then the class `Foo` would need to implement `ILogger<Foo>`, but that is not what one would want to do. What we would want to do is in any type to specify that we want an `ILogger` for the current type.

**Reply by jnm2** [reply]
I'm not seeing the difference in terms of use cases, only in terms of preferred syntax to achieve the same end. Is that accurate?

**Reply by Timo-Weike** [reply]
No, because it is not the same. 
For example, in the case of the `ILogger<T>` case, if you are in a ASP.Net core application and have a controller `Foo` and you inject an `ILogger<Foo>` into the controller, you will get an object of the type `Microsoft.Extensions.Logging.Logger<Foo>` which implements the interface `ILogger<Foo>`. But the proposal you linked proposes a feature which lets you specify that a generic parameter for an interface has to be the implementing type. 

But that is something you do not want for the `ILogger<T>` interface.

My proposed feature would only be a kind of syntactic suger, so you do not need to repeat the name of a type if it references itself in itself.

**Reply by jnm2** [reply]
Are you sure though? The proposal I linked would do this:

```cs
public class Foo<T> where T : this
{
    public Foo(ILogger<T> logger)
    {
    }
}
```

```cs
public class Foo<T> : IComparable<T>, IEquatable<T>
    where T : this
{
```

This should work everywhere the nearest declaring-type is clear, so even in a static method and neasted classes.

```cs
public class Foo<T> where T : this
{
    class Bar : IComparable<T> { ... }
}
```

Seems to me the only difference is in whether `Foo` is generic or not. Meeting notes for #5413 considered this difference and said:

> We also talked about the possibility of using associated types to represent the self type, rather than a public type parameter. While this is somewhat attractive, we're in general agreement that unless we have real runtime support for an associated self type, we don't want to take this direction. In C#, we generally avoid having metadata significantly differ from the source; if the metadata needs to have a generic type parameter, then we would like the source to indicate such. At some point, associated types need to be translated into the universal type space, which will spread virally throughout the program: that either needs to be done at compile time by Roslyn (the representation change we want to avoid), or it needs to be done by the runtime. The runtime has concerns about avoiding an MxN problem (where M methods will need to be specialized N times) without whole-program analysis, which is not always available.

(https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-11-10.md#self-types)

**Reply by Timo-Weike** [reply]
But this would not solve the problem, because if you have:
~~~ cs
public class Foo<T> where T : this
{
    public Foo(ILogger<T> logger)
    {
    }
}
~~~
the `T` could not be the type `Foo` but would need to be it's inheriting type, so you would need to do
~~~ cs
public class Bar : Foo<Bar>
{
    public Bar(ILogger<Bar> logger)
        : base(logger) { }
}
~~~
and now you haven't won a thing because you still have to write `ILogger<Bar>` and you could also not do `Foo<Foo>` then you run into an endless recursive generic type and would need to write `Foo<Foo<Foo<...>>`.

Furthermore, in the case of ASP.Net core: I think it is not possible to have a controller with a generic parameter or at least I would not know how to register it with the DI-container.

I understand an accept if my proposed feature is something that was discussed before and was agreed upon to not do, but I would not consider the too features equal, simply because I would not use them for the same goals.

**Comment by svick** [comment]
I dislike the suggested syntax: `typeof` in C# gives you a `Type`, but here it would give you a type. Those are two very different things and they should not be conflated.

**Reply by Timo-Weike** [reply]
I agree with you, I also am not happy with it. I also though about proposing using `this` but it collides with different features (e.g. [self-constraint](https://github.com/dotnet/csharplang/blob/main/proposals/self-constraint.md#self-constraint-for-generic-type-parameters)).
